File: android/app/src/main/java/com/translatehelper/TranslationCore.kt
package com.translatehelper

data class LanguageProfile(
    val code: String,
    val englishName: String,
    val chineseName: String,
)

data class TranslationResult(
    val sourceText: String,
    val translatedText: String,
    val sourceLang: String,
    val targetLang: String,
    val status: String,
    val fromCache: Boolean = false,
    val skipped: Boolean = false,
)

object TranslationCore {
    val supportedSourceLanguages = linkedMapOf(
        "en" to LanguageProfile("en", "English", "英语"),
        "ja" to LanguageProfile("ja", "Japanese", "日语"),
        "ko" to LanguageProfile("ko", "Korean", "韩语"),
        "vi" to LanguageProfile("vi", "Vietnamese", "越南语"),
    )

    val targetLanguage = LanguageProfile("zh-CN", "Chinese", "简体中文")
    val defaultSourceLangs = listOf("en", "ja", "ko", "vi")

    private val jaKanaRegex = Regex("[\\u3040-\\u309f\\u30a0-\\u30ff]")
    private val koHangulRegex = Regex("[\\uac00-\\ud7af]")
    private val viDiacriticRegex = Regex(
        "[àáảãạăằắẳẵặâầấẩẫậèéẻẽẹêềếểễệìíỉĩịòóỏõọôồốổỗộơờớởỡợùúủũụưừứửữựỳýỷỹỵđ]",
        RegexOption.IGNORE_CASE,
    )
    private val cjkRegex = Regex("[\\u4e00-\\u9fff]")
    private val latinRegex = Regex("[A-Za-z]")

    fun parseSourceLangs(raw: String?): List<String> {
        if (raw.isNullOrBlank()) {
            return defaultSourceLangs
        }
        val codes = mutableListOf<String>()
        raw.split(",").forEach { part ->
            val code = part.trim().lowercase()
            if (code.isNotEmpty() && supportedSourceLanguages.containsKey(code) && !codes.contains(code)) {
                codes.add(code)
            }
        }
        return if (codes.isEmpty()) defaultSourceLangs else codes
    }

    fun languageLabel(code: String, chinese: Boolean = true): String {
        if (code == targetLanguage.code) {
            return if (chinese) targetLanguage.chineseName else targetLanguage.englishName
        }
        val profile = supportedSourceLanguages[code] ?: return code
        return if (chinese) profile.chineseName else profile.englishName
    }

    fun formatRouteLabel(sourceLang: String, targetLang: String = targetLanguage.code): String {
        return "${languageLabel(sourceLang)} → ${languageLabel(targetLang)}"
    }

    fun detectSourceLanguage(text: String, allowed: List<String> = defaultSourceLangs): String {
        val sample = text.trim()
        if (sample.isEmpty()) {
            return "unknown"
        }
        if (koHangulRegex.containsMatchIn(sample) && "ko" in allowed) {
            return "ko"
        }
        if (jaKanaRegex.containsMatchIn(sample) && "ja" in allowed) {
            return "ja"
        }
        if (viDiacriticRegex.containsMatchIn(sample) && "vi" in allowed) {
            return "vi"
        }

        val cjkCount = cjkRegex.findAll(sample).count()
        val latinCount = latinRegex.findAll(sample).count()
        if (cjkCount > 0 && latinCount == 0 &&
            !jaKanaRegex.containsMatchIn(sample) &&
            !koHangulRegex.containsMatchIn(sample)
        ) {
            return targetLanguage.code
        }
        if (latinCount > 0) {
            if ("vi" in allowed && viDiacriticRegex.containsMatchIn(sample)) {
                return "vi"
            }
            if ("en" in allowed) {
                return "en"
            }
        }
        return "unknown"
    }

    fun shouldSkipTranslation(
        text: String,
        detectedLang: String,
        targetLang: String = targetLanguage.code,
    ): Pair<Boolean, String> {
        if (detectedLang == targetLang) {
            return true to "已是中文，无需翻译"
        }
        if (text.isBlank()) {
            return true to "文本为空"
        }
        return false to ""
    }

    fun buildSystemPrompt(
        sourceLang: String,
        targetLang: String = targetLanguage.code,
        allowedSources: List<String> = defaultSourceLangs,
    ): String {
        val allowedNames = allowedSources.joinToString("、") { languageLabel(it) }
        val targetName = languageLabel(targetLang)
        val sourceHint = when (sourceLang) {
            "unknown" -> "自动识别源语言，优先支持：$allowedNames。若文本已是中文则原样返回。"
            targetLanguage.code -> "源语言是中文，直接返回原文，不要翻译。"
            else -> "源语言是${languageLabel(sourceLang)}（$sourceLang）。"
        }
        return """
            你是专业翻译引擎。
            $sourceHint
            将用户文本翻译为$targetName（$targetLang）。
            要求：
            1. 只输出译文，不要解释、不要前缀。
            2. 保留原文中的专有名词、数字、URL、代码片段。
            3. 译文使用自然流畅的简体中文。
        """.trimIndent()
    }

    fun buildTranslationResult(
        sourceText: String,
        translatedText: String,
        sourceLang: String,
        targetLang: String = targetLanguage.code,
        fromCache: Boolean = false,
        skipped: Boolean = false,
    ): TranslationResult {
        val route = formatRouteLabel(sourceLang, targetLang)
        val status = when {
            skipped -> "已是中文"
            fromCache -> "$route（缓存）"
            else -> route
        }
        return TranslationResult(
            sourceText = sourceText,
            translatedText = translatedText,
            sourceLang = sourceLang,
            targetLang = targetLang,
            status = status,
            fromCache = fromCache,
            skipped = skipped,
        )
    }
}

File: shared/translation_core.py
"""Shared translation logic for desktop and future mobile clients."""

from __future__ import annotations

import re
from dataclasses import dataclass


@dataclass(frozen=True)
class LanguageProfile:
    code: str
    english_name: str
    chinese_name: str


SUPPORTED_SOURCE_LANGUAGES: dict[str, LanguageProfile] = {
    "en": LanguageProfile("en", "English", "英语"),
    "ja": LanguageProfile("ja", "Japanese", "日语"),
    "ko": LanguageProfile("ko", "Korean", "韩语"),
    "vi": LanguageProfile("vi", "Vietnamese", "越南语"),
}

TARGET_LANGUAGE = LanguageProfile("zh-CN", "Chinese", "简体中文")

DEFAULT_SOURCE_LANGS = ("en", "ja", "ko", "vi")

_JA_KANA_RE = re.compile(r"[\u3040-\u309f\u30a0-\u30ff]")
_KO_HANGUL_RE = re.compile(r"[\uac00-\ud7af]")
_VI_DIACRITIC_RE = re.compile(
    r"[àáảãạăằắẳẵặâầấẩẫậèéẻẽẹêềếểễệìíỉĩịòóỏõọôồốổỗộơờớởỡợùúủũụưừứửữựỳýỷỹỵđ]",
    re.IGNORECASE,
)
_CJK_RE = re.compile(r"[\u4e00-\u9fff]")
_LATIN_RE = re.compile(r"[A-Za-z]")


@dataclass(frozen=True)
class TranslationResult:
    source_text: str
    translated_text: str
    source_lang: str
    target_lang: str
    status: str
    from_cache: bool = False
    skipped: bool = False


def parse_source_langs(raw: str | None) -> tuple[str, ...]:
    if not raw or not raw.strip():
        return DEFAULT_SOURCE_LANGS

    codes: list[str] = []
    for part in raw.split(","):
        code = part.strip().lower()
        if not code:
            continue
        if code not in SUPPORTED_SOURCE_LANGUAGES:
            continue
        if code not in codes:
            codes.append(code)
    return tuple(codes) if codes else DEFAULT_SOURCE_LANGS


def language_label(code: str, *, chinese: bool = True) -> str:
    if code == TARGET_LANGUAGE.code:
        return TARGET_LANGUAGE.chinese_name if chinese else TARGET_LANGUAGE.english_name
    profile = SUPPORTED_SOURCE_LANGUAGES.get(code)
    if profile is None:
        return code
    return profile.chinese_name if chinese else profile.english_name


def format_route_label(source_lang: str, target_lang: str = TARGET_LANGUAGE.code) -> str:
    return f"{language_label(source_lang)} → {language_label(target_lang)}"


def detect_source_language(text: str, allowed: tuple[str, ...] = DEFAULT_SOURCE_LANGS) -> str:
    sample = text.strip()
    if not sample:
        return "unknown"

    if _KO_HANGUL_RE.search(sample) and "ko" in allowed:
        return "ko"
    if _JA_KANA_RE.search(sample) and "ja" in allowed:
        return "ja"
    if _VI_DIACRITIC_RE.search(sample) and "vi" in allowed:
        return "vi"

    cjk_count = len(_CJK_RE.findall(sample))
    latin_count = len(_LATIN_RE.findall(sample))

    if cjk_count > 0 and latin_count == 0 and not _JA_KANA_RE.search(sample) and not _KO_HANGUL_RE.search(sample):
        return TARGET_LANGUAGE.code

    if latin_count > 0:
        if "vi" in allowed and _VI_DIACRITIC_RE.search(sample):
            return "vi"
        if "en" in allowed:
            return "en"

    return "unknown"


def should_skip_translation(text: str, detected_lang: str, target_lang: str = TARGET_LANGUAGE.code) -> tuple[bool, str]:
    if detected_lang == target_lang:
        return True, "已是中文，无需翻译"
    if not text.strip():
        return True, "文本为空"
    return False, ""


def build_system_prompt(
    *,
    source_lang: str,
    target_lang: str,
    allowed_sources: tuple[str, ...] = DEFAULT_SOURCE_LANGS,
) -> str:
    allowed_names = "、".join(language_label(code) for code in allowed_sources)
    target_name = language_label(target_lang)

    if source_lang == "unknown":
        source_hint = (
            f"自动识别源语言，优先支持：{allowed_names}。"
            "若文本已是中文则原样返回。"
        )
    elif source_lang == TARGET_LANGUAGE.code:
        source_hint = "源语言是中文，直接返回原文，不要翻译。"
    else:
        source_hint = f"源语言是{language_label(source_lang)}（{source_lang}）。"

    return (
        "你是专业翻译引擎。\n"
        f"{source_hint}\n"
        f"将用户文本翻译为{target_name}（{target_lang}）。\n"
        "要求：\n"
        "1. 只输出译文，不要解释、不要前缀。\n"
        "2. 保留原文中的专有名词、数字、URL、代码片段。\n"
        "3. 译文使用自然流畅的简体中文。"
    )


def build_translation_result(
    *,
    source_text: str,
    translated_text: str,
    source_lang: str,
    target_lang: str,
    from_cache: bool = False,
    skipped: bool = False,
) -> TranslationResult:
    route = format_route_label(source_lang, target_lang)
    if skipped:
        status = "已是中文"
    elif from_cache:
        status = f"{route}（缓存）"
    else:
        status = route
    return TranslationResult(
        source_text=source_text,
        translated_text=translated_text,
        source_lang=source_lang,
        target_lang=target_lang,
        status=status,
        from_cache=from_cache,
        skipped=skipped,
    )
